File: movil/lib/core/catalogo/catalogo_admin_service.dart
import '../api.dart';
import 'catalogo_admin_models.dart';

/// CU10: Gestionar Catalogo (prendas, variantes, imagenes, categorias y marcas).
class CatalogoAdminService {
  Future<ReferenciasOut> obtenerReferencias() async {
    final respuesta = await api.get('/admin/catalogo/referencias');
    return ReferenciasOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<List<ProductoAdminOut>> listarProductos({
    String? q,
    String? categoriaId,
    bool? activo,
  }) async {
    final respuesta = await api.get('/admin/catalogo/productos', query: {
      'q': q,
      'categoria_id': categoriaId,
      'activo': activo,
    });
    return comoLista(respuesta).map(ProductoAdminOut.desdeJson).toList();
  }

  Future<ProductoAdminDetalleOut> obtenerProducto(String productoId) async {
    final respuesta = await api.get('/admin/catalogo/productos/$productoId');
    return ProductoAdminDetalleOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<ProductoAdminDetalleOut> crearProducto(ProductoIn datos) async {
    final respuesta = await api.post('/admin/catalogo/productos', cuerpo: datos.aJson());
    return ProductoAdminDetalleOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<ProductoAdminDetalleOut> actualizarProducto(String productoId, ProductoIn datos) async {
    final respuesta =
        await api.put('/admin/catalogo/productos/$productoId', cuerpo: datos.aJson());
    return ProductoAdminDetalleOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<ProductoAdminDetalleOut> cambiarEstadoProducto(String productoId, bool activo) async {
    final respuesta = await api
        .patch('/admin/catalogo/productos/$productoId/estado', cuerpo: {'activo': activo});
    return ProductoAdminDetalleOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<VarianteAdminOut> crearVariante(String productoId, VarianteIn datos) async {
    final respuesta = await api.post(
      '/admin/catalogo/productos/$productoId/variantes',
      cuerpo: datos.aJson(),
    );
    return VarianteAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<VarianteAdminOut> actualizarVariante(String varianteId, VarianteIn datos) async {
    final respuesta =
        await api.put('/admin/catalogo/variantes/$varianteId', cuerpo: datos.aJson());
    return VarianteAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<VarianteAdminOut> cambiarEstadoVariante(String varianteId, bool activa) async {
    final respuesta = await api
        .patch('/admin/catalogo/variantes/$varianteId/estado', cuerpo: {'activo': activa});
    return VarianteAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<ImagenOut> agregarImagen(
    String productoId, {
    required String url,
    String uso = 'CATALOGO',
    int? colorId,
    bool esPrincipal = false,
    int orden = 0,
  }) async {
    final respuesta = await api.post('/admin/catalogo/productos/$productoId/imagenes', cuerpo: {
      'url': url,
      'uso': uso,
      'color_id': colorId,
      'es_principal': esPrincipal,
      'orden': orden,
    });
    return ImagenOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<void> eliminarImagen(String imagenId) =>
      api.delete('/admin/catalogo/imagenes/$imagenId');

  Future<List<CategoriaAdminOut>> listarCategorias() async {
    final respuesta = await api.get('/admin/catalogo/categorias');
    return comoLista(respuesta).map(CategoriaAdminOut.desdeJson).toList();
  }

  Future<CategoriaAdminOut> crearCategoria({
    required String nombre,
    String? categoriaPadreId,
    String? imagenUrl,
    int orden = 0,
  }) async {
    final respuesta = await api.post('/admin/catalogo/categorias', cuerpo: {
      'nombre': nombre,
      'categoria_padre_id': categoriaPadreId,
      'imagen_url': (imagenUrl?.isEmpty ?? true) ? null : imagenUrl,
      'orden': orden,
    });
    return CategoriaAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<CategoriaAdminOut> actualizarCategoria(
    String categoriaId, {
    required String nombre,
    String? categoriaPadreId,
    String? imagenUrl,
    int orden = 0,
  }) async {
    final respuesta = await api.put('/admin/catalogo/categorias/$categoriaId', cuerpo: {
      'nombre': nombre,
      'categoria_padre_id': categoriaPadreId,
      'imagen_url': (imagenUrl?.isEmpty ?? true) ? null : imagenUrl,
      'orden': orden,
    });
    return CategoriaAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<CategoriaAdminOut> cambiarEstadoCategoria(String categoriaId, bool activa) async {
    final respuesta = await api
        .patch('/admin/catalogo/categorias/$categoriaId/estado', cuerpo: {'activa': activa});
    return CategoriaAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<MarcaAdminOut> crearMarca(String nombre, {String? logoUrl}) async {
    final respuesta = await api.post('/admin/catalogo/marcas', cuerpo: {
      'nombre': nombre,
      'logo_url': (logoUrl?.isEmpty ?? true) ? null : logoUrl,
    });
    return MarcaAdminOut.desdeJson(respuesta as Map<String, dynamic>);
  }
}

final catalogoAdminService = CatalogoAdminService();

File: frontend/src/app/core/catalogo/catalogo-admin.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import {
  CategoriaAdminOut,
  CategoriaIn,
  ImagenIn,
  ImagenOut,
  MarcaAdminOut,
  ProductoAdminDetalleOut,
  ProductoAdminOut,
  ProductoIn,
  PromocionAdminOut,
  PromocionIn,
  ReferenciasOut,
  VarianteAdminOut,
  VarianteIn,
} from './catalogo-admin.models';

/** CU10 - Gestionar Catalogo. */
@Injectable({ providedIn: 'root' })
export class CatalogoAdminService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/admin/catalogo`;

  referencias(): Observable<ReferenciasOut> {
    return this.http.get<ReferenciasOut>(`${this.base}/referencias`);
  }

  listarProductos(
    q?: string | null,
    categoriaId?: string | null,
    activo?: boolean | null,
  ): Observable<ProductoAdminOut[]> {
    let params = new HttpParams();
    if (q) params = params.set('q', q);
    if (categoriaId) params = params.set('categoria_id', categoriaId);
    if (activo !== null && activo !== undefined) params = params.set('activo', activo);
    return this.http.get<ProductoAdminOut[]>(`${this.base}/productos`, { params });
  }

  obtenerProducto(id: string): Observable<ProductoAdminDetalleOut> {
    return this.http.get<ProductoAdminDetalleOut>(`${this.base}/productos/${id}`);
  }

  crearProducto(datos: ProductoIn): Observable<ProductoAdminDetalleOut> {
    return this.http.post<ProductoAdminDetalleOut>(`${this.base}/productos`, datos);
  }

  actualizarProducto(id: string, datos: ProductoIn): Observable<ProductoAdminDetalleOut> {
    return this.http.put<ProductoAdminDetalleOut>(`${this.base}/productos/${id}`, datos);
  }

  cambiarEstadoProducto(id: string, activo: boolean): Observable<ProductoAdminDetalleOut> {
    return this.http.patch<ProductoAdminDetalleOut>(`${this.base}/productos/${id}/estado`, { activo });
  }

  crearVariante(productoId: string, datos: VarianteIn): Observable<VarianteAdminOut> {
    return this.http.post<VarianteAdminOut>(`${this.base}/productos/${productoId}/variantes`, datos);
  }

  actualizarVariante(varianteId: string, datos: VarianteIn): Observable<VarianteAdminOut> {
    return this.http.put<VarianteAdminOut>(`${this.base}/variantes/${varianteId}`, datos);
  }

  cambiarEstadoVariante(varianteId: string, activo: boolean): Observable<VarianteAdminOut> {
    return this.http.patch<VarianteAdminOut>(`${this.base}/variantes/${varianteId}/estado`, { activo });
  }

  agregarImagen(productoId: string, datos: ImagenIn): Observable<ImagenOut> {
    return this.http.post<ImagenOut>(`${this.base}/productos/${productoId}/imagenes`, datos);
  }

  subirImagen(
    productoId: string,
    archivo: File,
    opciones: { uso: string; esPrincipal: boolean; orden: number },
  ): Observable<ImagenOut> {
    const formData = new FormData();
    formData.append('archivo', archivo);
    formData.append('uso', opciones.uso);
    formData.append('es_principal', String(opciones.esPrincipal));
    formData.append('orden', String(opciones.orden));
    return this.http.post<ImagenOut>(`${this.base}/productos/${productoId}/imagenes/subir`, formData);
  }

  eliminarImagen(imagenId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/imagenes/${imagenId}`);
  }

  listarCategorias(): Observable<CategoriaAdminOut[]> {
    return this.http.get<CategoriaAdminOut[]>(`${this.base}/categorias`);
  }

  crearCategoria(datos: CategoriaIn): Observable<CategoriaAdminOut> {
    return this.http.post<CategoriaAdminOut>(`${this.base}/categorias`, datos);
  }

  actualizarCategoria(id: string, datos: CategoriaIn): Observable<CategoriaAdminOut> {
    return this.http.put<CategoriaAdminOut>(`${this.base}/categorias/${id}`, datos);
  }

  cambiarEstadoCategoria(id: string, activa: boolean): Observable<CategoriaAdminOut> {
    return this.http.patch<CategoriaAdminOut>(`${this.base}/categorias/${id}/estado`, { activa });
  }

  crearMarca(nombre: string, logoUrl: string | null): Observable<MarcaAdminOut> {
    return this.http.post<MarcaAdminOut>(`${this.base}/marcas`, { nombre, logo_url: logoUrl });
  }

  listarPromociones(activa?: boolean | null): Observable<PromocionAdminOut[]> {
    let params = new HttpParams();
    if (activa !== null && activa !== undefined) params = params.set('activa', activa);
    return this.http.get<PromocionAdminOut[]>(`${this.base}/promociones`, { params });
  }

  crearPromocion(datos: PromocionIn): Observable<PromocionAdminOut> {
    return this.http.post<PromocionAdminOut>(`${this.base}/promociones`, datos);
  }

  actualizarPromocion(id: string, datos: PromocionIn): Observable<PromocionAdminOut> {
    return this.http.put<PromocionAdminOut>(`${this.base}/promociones/${id}`, datos);
  }

  cambiarEstadoPromocion(id: string, activa: boolean): Observable<PromocionAdminOut> {
    return this.http.patch<PromocionAdminOut>(`${this.base}/promociones/${id}/estado`, { activa });
  }
}
